### PR TITLE
Optimize layout of function arguments in the Rust ABI - take 2

### LINKED ISSUE #97540: `Result` that uses niches resulting in a final size of 16 bytes emits poor LLVM IR due to ABI
The following code:
```rust
pub fn bad() -> Result<u64, core::ptr::NonNull<()>> {
    Ok(0)
}

#[allow(improper_ctypes_definitions)]
pub extern "C" fn good() -> Result<u64, core::ptr::NonNull<()>> {
    Ok(0)
}
```
[godbolt LLVM IR](https://godbolt.org/z/b5jjv43nK)
[godbolt ASM](https://godbolt.org/z/jKM8j6hEn)

Both functions should be identical, returning the 128-bit value in RAX and RDX per the sysv-64 ABI, which is the default for the environment godbolt uses by default.

Instead, when returning the value without specifying any ABI, the compiler chooses to return the value using an out pointer argument.

**Note:** This affects `Result<usize, std::io::Error>`, which is a particularly useful type and affects a significant portion of `std::io`, such as `Read` and `Write`

### Meta

This has improved and regressed since 1.43.0, the first version I could find which would use 128 bits for the type.
All versions tested used the expected code generation for the `extern "C"` function.
All tests done using https://godbolt.org and no target overrides, which in practice is a x86_64 linux machine using the sysv-64 ABI.

..= 1.42.0: niche not used, cannot be compared
1.43.0 ..= 1.47.0: codegen like today, uses out ptr for return
1.48.0 ..= 1.60.0: returns using an LLVM `i128`, which produces the most desirable code on sysv-64
1.61.0 .. (current 2022-05-28 nightly): poor codegen that uses an out ptr for return

@rustbot label +regression-from-stable-to-stable +A-codegen

### PR DESCRIPTION
**Pull request by Urgau**
This PR is another attempt to solving the problem of floats being aggregated as integer, it's purpose is to supersede https://github.com/rust-lang/rust/pull/94570 which was a big bulldozer and has cause some problem like https://github.com/rust-lang/rust/issues/97540. 

This PR is meant to address issues like https://github.com/rust-lang/rust/issues/85265#issuecomment-985369739 where floats where being aggregated as integer causing poor codegen. It does this by only passing homogeneous aggregate of floats as array if <= ptr-size and by pointer on array if > ptr-size as an optimization.

It's also a reopening of https://github.com/rust-lang/rust/pull/93564 with hopefully no regression at all.

<details>

<summary>
Rust code + asm before/after
</summary>

Rust code:
```rust
pub fn foo() -> Result<u64, core::ptr::NonNull<()>> {
    Ok(0)
}
```

ASM Before:
```asm
foo:
        mov     rax, rdi
        xorps   xmm0, xmm0
        movups  xmmword ptr [rdi], xmm0
        ret
```

ASM After:
```asm
foo:
	xor	eax, eax
	xor	edx, edx
	ret
```

</details>

cc @nbdd0121 @workingjubilee

Revert https://github.com/rust-lang/rust/pull/94570
Fixes https://github.com/rust-lang/rust/issues/97540

### CODE REVIEW

File: compiler/rustc_middle/src/ty/layout.rs
Comment (nbdd0121) [inline compiler/rustc_middle/src/ty/layout.rs:3388]: As I pointed out in https://github.com/rust-lang/rust/issues/97540#issuecomment-1141411742, this check should be done before further examing the layout.

Comment (Urgau) [reply compiler/rustc_middle/src/ty/layout.rs:3388]: It would be my pleasure but I don't see how to reverse the conditions; if you see how, please tell me.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (nbdd0121) [inline compiler/rustc_middle/src/ty/layout.rs:3391]:
```suggestion
                        } else if size > Pointer.size(self) && unlikely!(self.has_all_float(&arg.layout)) {
                            // We don't want to aggregate floats as an aggregates of Integer
                            // because this will hurt the generated assembly (#93490)
                            //
                            // As an optimization we want to pass homogeneous aggregate of floats
                            // greater than pointer size as indirect
                            arg.make_indirect();
                        } else {
```

Comment (Urgau) [reply compiler/rustc_middle/src/ty/layout.rs:3391]: This is not the same logic as before. Before your suggestion if it was `self.has_all_float(&arg.layout)` but NOT `size > Pointer.size(self)` nothing would happen, with this suggestion it will take the else branch doing exactly the opposite of what we want to make.

Comment (nbdd0121) [reply compiler/rustc_middle/src/ty/layout.rs:3391]: Also, even if it's desirable that small aggregates with float has a special treatment, it should probably be a separate PR that's backed by separate perf run.

Comment (Urgau) [reply compiler/rustc_middle/src/ty/layout.rs:3391]: Well on `x86_64` this would mean that `f64` -> `i64`, `f32` -> `i32` and `[f32; 2]` -> `i64` which is not ideal and is the reason of this PR, ie let the possibility for the codegen to use floating point register for small aggregate of floats or to put them behind an indirection.

Comment (nbdd0121) [reply compiler/rustc_middle/src/ty/layout.rs:3391]: `f64` and `f32` are not `Abi::Aggregate` and will not hit this path. For `[f32; 2]`, the current status quo is to convert it to integer.

Essentially this PR is doing two things:
1. Pass not-all-float aggregates smaller than 2 register directly
2. Stop casting all-float aggregates smaller than 1 register to integer

I am suggesting just do 1 in this PR.

Comment (Urgau) [reply compiler/rustc_middle/src/ty/layout.rs:3391]: I tried your suggestion locally and only one test failed. I will push it after the perf run so we can have a baseline to know if it's even necessary or not.

Comment (Urgau) [reply compiler/rustc_middle/src/ty/layout.rs]: Well the perf impact of introducing this "fast path" is negligible: https://github.com/rust-lang/rust/pull/97559#issuecomment-1142418639

**Comment by apiraino** [comment]
Maybe ready for another round of review after the latest perf. run? Please switch back to the author if more work is needed, thanks!

@rustbot review

**Review by workingjubilee — CHANGES_REQUESTED** [review]
r? @workingjubilee

This is not an improvement in compile times, and I do not think the improvement in codegen is so clearly significant, from the examples given, as to warrant merging in spite of that. It would be sufficient if we had a sample microbenchmark this wins on in terms of runtime before/after.

@rustbot author

**Comment by workingjubilee** [comment]
Well also to be honest I will take "anything that shows that we actually have a solid win here", actually, such as removing the compiler perf regression, or a more complex codegen example involving multiple interacting functions that gets simplified significantly by this.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (workingjubilee) [inline compiler/rustc_middle/src/ty/layout.rs:3383]: Curious: Why the `unlikely!` hint?

**Comment by Urgau** [comment]
> Maybe ready for another round of review after the latest perf. run? Please switch back to the author if more work is needed, thanks!

This is PR indeed needs more work but I was voluntarily waiting on @cjgillot review to guide me on a possible solution(s) to reduce the impact of my changes. Any help @workingjubilee ?

> This is not an improvement in compile times, and I do not think the improvement in codegen is so clearly significant, from the examples given, as to warrant merging in spite of that.

The reasoning is the same as https://github.com/rust-lang/rust/pull/94570 except that this PR makes an exception for floats so they don't get aggregated at all and increase the threshold for aggregating in general (from ptr-size to ptr-size*2). 

>  It would be sufficient if we had a sample microbenchmark this wins on in terms of runtime before/after.

If only rustc-perf would be able to bench runtime execution instead of compile time execution.
Anything in particular you would want ? Which tool do you want me to use ?

File: compiler/rustc_middle/src/ty/layout.rs
Comment (workingjubilee) [reply compiler/rustc_middle/src/ty/layout.rs:3383]: `unlikely!` hints can result in the function getting outlined even when inlining it would be profitable, thereby adding an indirection and making the call even slower (the first call, for instance, will always look a certain way).

**Comment by workingjubilee** [comment]
> If only rustc-perf would be able to bench runtime execution instead of compile time execution.

I know, right?

This is a really tricky problem, and I don't believe adding tiny little bits of conditional logic is going to help it that much. Ideally, the goal is to make the logic more **unconditional**, instead of carefully mangling conditions. CPUs love it when they can just rush through a task and do the same steps each time. Unless they can predict the branch each time, it's often better to do four things unconditionally than add a branch to select between doing three things or one thing.

So in that regard, I believe making special exceptions for floats in particular may be part of what is misguided in our approaches, as the impact of not autovectorizing in certain cases may be heavier on floats, but it's not like we're exactly winning if we whiff on integers also.

**Comment by Kobzol** [comment]
> If only rustc-perf would be able to bench runtime execution instead of compile time execution.

FWIW, we're actively thinking about this and hopefully it will be possible eventually. But not in the immediate future.

**Comment by nbdd0121** [comment]
I wonder if we should determine this through number of fields. For example, if this is a `[f32 x 4]`, it has 4 fields which probably should be passed indirectly. For a `Result<isize, usize>` it only has 2 (nonoverlapping) fields so could be passed directly.

I also wonder if we can remove the cast-to-integer behaviour and let LLVM deal with small aggregates. The comment justifying the behaviour is introduced in #32080 in 2016 and a lot must have been changed.

**Comment by workingjubilee** [comment]
Those both sound like profitable directions!
Though I suggest not trying to do both at once, @Urgau, for the scientific reason of observing both hypotheses independently first before trying to combine them.

**Comment by mati865** [comment]
> If only rustc-perf would be able to bench runtime execution instead of compile time execution.

You can take a look at https://github.com/anp/lolbench
I had used it long time ago for different PR but it involved some hacking in source code.

**Comment by Urgau** [comment]
> I also wonder if we can remove the cast-to-integer behaviour and let LLVM deal with small aggregates. The comment justifying the behaviour is introduced in #32080 in 2016 and a lot must have been changed.

Let's try this. I just pushed a change to remove the small-aggregate casting to Integer.
I didn't update (yet) the tests but it compiles; could someone starts an perf run.

**Comment by Urgau** [comment]
Well, this appears to be a miscompilation of `rustc` itself.
I was also told that relying on LLVM algo for aggregate type is not a good idea.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (eddyb) [inline compiler/rustc_middle/src/ty/layout.rs:3399]: Why not use the "homogeneous aggregate" concept that `rustc_target::abi::call` has?
(Unless you want to react to *any* floats being present in the aggregate, instead of it being *all* floats)

**Comment by eddyb** [comment]
> > I also wonder if we can remove the cast-to-integer behaviour and let LLVM deal with small aggregates. The comment justifying the behaviour is introduced in #32080 in 2016 and a lot must have been changed.
> 
> Let's try this. I just pushed a change to remove the small-aggregate casting to Integer. I didn't update (yet) the tests but it compiles; could someone starts an perf run.

"Just let LLVM deal with it" isn't a real option, and we hope to make it impossible eventually (see #98615) - that is, one shouldn't expect there is a "LLVM (aggregate) type" for every Rust type.

Part of the reason is that these types may be as bad as `{ [i8 x sizeof(T)] }`, which serve absolutely no purpose other than passing the size of the type to LLVM in contexts where LLVM doesn't offer a more direct way to describe the size.

When passed "by value" in LLVM's "first-class aggregate" semantics, they will likely consume one register per `i8` component (until they start spilling into the stack).
And that will happen even for such `[i8 x _]` arrays we insert *for padding*, as LLVM has no functionality for describing aggregate layout (though I'm not saying they should add some, #98615 contains some of my commentary on the matter, i.e. I believe they should just remove aggregates).

`abi::call::PassMode::Cast` prevents such issues by describing how to reinterpret the value as appropriately-sized registers - once we get around to removing most type distinctions in backends (other than bool/int/float/opaque-ptr, and vectors thereof), it should become the primary way for describing direct value passing (this also goes well in hand with some of my ancient plans of incorporating more register/stack knowledge into `rustc_target::abi::call`, which could aid backends like cranelift, or be used to extensively test our understanding of how LLVM abstracts these things, etc.).

If you want to see what it takes to optimize which registers are used, you should take a look at e.g. `rustc_target::abi::call::x86_64` and how it recurses into aggregates and processes leaves by offset.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (Urgau) [reply compiler/rustc_middle/src/ty/layout.rs:3399]: If you're talking about [`rustc_target::abi::call::HomogeneousAggregate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/call/enum.HomogeneousAggregate.html) I already tried it in https://github.com/rust-lang/rust/pull/93564 and it gave some potentially non-negliable regressions albeit it was only in one crate, maybe things have improved things then. Do you want me to retry with it ?

**Comment by nbdd0121** [comment]
My issue with the current casting behaviour is that it forces everything through integer registers, which is really suboptimal when float is involved. My hope was that by using LLVM aggregates, LLVM could generate more optimal code, but obviously that's not case. I suppose another way would be having some logic that does classificiation for the Rust ABI similar to how x86_64 ABI classifies arguments.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (Urgau) [reply compiler/rustc_middle/src/ty/layout.rs:3399]: "Any" was not what I was targeting here, but it's a thing I also want to target at some point.

**Comment by Urgau** [comment]
Let's retry with the [`rustc_target::abi::call::HomogeneousAggregate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/call/enum.HomogeneousAggregate.html) concept. As suggested by @eddyb in https://github.com/rust-lang/rust/pull/97559#discussion_r912461652.

**Comment by Urgau** [comment]
I have done some benchmarks with hyperfine my stage1, nightly, stable, the help of https://github.com/mariomka/regex-benchmark and it showed a consistent discrease in performance in debug which match the regression on the compile-time. Seems that aggregating floats into integer is better (or I'm doing something wrong ?). What should be the next step ?

```bash
$ hyperfine -N --warmup 3 -n nightly "./target_nightly/debug/rg '[\\w]+://[^/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?' ./input-text.txt" -n stage1 "./target_stage1/debug/rg '[\\w]+://[^/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?' ./input-text.txt"
Benchmark 1: nightly
  Time (mean ± σ):     400.1 ms ±  11.8 ms    [User: 395.1 ms, System: 3.2 ms]
  Range (min … max):   384.5 ms … 420.9 ms    10 runs

Benchmark 2: stage1
  Time (mean ± σ):     665.0 ms ±  14.3 ms    [User: 655.7 ms, System: 5.6 ms]
  Range (min … max):   642.8 ms … 686.7 ms    10 runs

Summary
  'nightly' ran
    1.66 ± 0.06 times faster than 'stage1'
```
```bash
$ hyperfine -N --warmup 3 -n nightly_release "./target_nightly_release/release/rg '[\\w]+://[^/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?' ./input-text.txt" -n stage1_release "./target_stage1_release/release/rg '[\\w]+://[^/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?' ./input-text.txt"
Benchmark 1: nightly_release
  Time (mean ± σ):      23.0 ms ±   2.7 ms    [User: 20.1 ms, System: 2.5 ms]
  Range (min … max):    16.7 ms …  30.9 ms    122 runs

Benchmark 2: stage1_release
  Time (mean ± σ):      23.0 ms ±   3.4 ms    [User: 20.1 ms, System: 2.6 ms]
  Range (min … max):    15.7 ms …  38.3 ms    122 runs

Summary
  'nightly_release' ran
    1.00 ± 0.19 times faster than 'stage1_release'
```
```bash
$ hyperfine -N --warmup 3 -n rust1_62_release "./target_rust1_62_release/release/rg '[\\w]+://[^/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?' ./input-text.txt" -n stage1_release "./target_stage1_release/release/rg '[\\w]+://[^/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?' ./input-text.txt"
Benchmark 1: rust1_62_release
  Time (mean ± σ):      26.4 ms ±   2.4 ms    [User: 22.9 ms, System: 3.1 ms]
  Range (min … max):    21.7 ms …  34.2 ms    104 runs

Benchmark 2: stage1_release
  Time (mean ± σ):      26.6 ms ±   2.9 ms    [User: 23.2 ms, System: 3.0 ms]
  Range (min … max):    22.5 ms …  45.2 ms    109 runs

Summary
  'rust1_62_release' ran
    1.01 ± 0.14 times faster than 'stage1_release'
```
```bash
$ hyperfine -N --warmup 3 -n rust1_62 "./target_rust1_62/debug/rg '[\\w]+://[^/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?' ./input-text.txt" -n stage1 "./target_stage1/debug/rg '[\\w]+://[^/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?' ./input-text.txt"
Benchmark 1: rust1_62
  Time (mean ± σ):     738.0 ms ±  19.3 ms    [User: 728.8 ms, System: 4.3 ms]
  Range (min … max):   706.0 ms … 772.8 ms    10 runs

Benchmark 2: stage1
  Time (mean ± σ):     787.3 ms ±  18.3 ms    [User: 776.9 ms, System: 3.6 ms]
  Range (min … max):   767.8 ms … 813.4 ms    10 runs

Summary
  'rust1_62' ran
    1.07 ± 0.04 times faster than 'stage1'
```

**Comment by the8472** [comment]
I'm surprised that regex-processing would be affected by changes to float codegen at all.

File: compiler/rustc_middle/src/ty/layout.rs
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/layout.rs:3372]: The old code used `size > ptr_size`. You changed it to `size > ptr_size * 2`.

Comment (Urgau) [reply compiler/rustc_middle/src/ty/layout.rs:3372]: Oops, you're right I forgot about this. It explain the performance regressions.
I will revert this change. Thanks for finding this out.

**Comment by Urgau** [comment]
> I'm surprised that regex-processing would be affected by changes to float codegen at all.

Me too, and thanks to @bjorn3 https://github.com/rust-lang/rust/pull/97559#discussion_r913577542 we now know that it wasn't related to float at all but a changed that I done the the `max_by_val_size`, which explain the regressions in regex.
Could I get another perf run, it should now measure only the cost of not aggregating floats.

**Comment by eddyb** [comment]
> My issue with the current casting behaviour is that it forces everything through integer registers, which is really suboptimal when float is involved. My hope was that by using LLVM aggregates, LLVM could generate more optimal code, but obviously that's not case. I suppose another way would be having some logic that does classificiation for the Rust ABI similar to how x86_64 ABI classifies arguments.

@nbdd0121 Right, what SysV x64 ABI does is more or less the only way to reliably do this.

The whole thing with LLVM aggregates and us calling this mode "casting" obscures the fact that if you want to choose which registers something gets passed in, there's no way around looking at its contents and making a decision, nothing else can make that decision for you (though we can provide helpers for common patterns across ABIs, like the "homogeneous aggregate" one).

It would be much clearer if we were decomposing a Rust argument into LLVM scalar/vector arguments, and optionally a stack passing argument (but since Clang doesn't already do this for its ABI handling, we risk ABI incompatibilities, so we'd need a lot more testing - see also #65111).

**Comment by Urgau** [comment]
I don't currently have the time to continue this particular PR, especially if I want to follow [eddyb comment](https://github.com/rust-lang/rust/pull/97559#issuecomment-1177333116).

Closing it.

**Pull request closed without merging**
